### ISSUE TITLE
Circuit Visualization

### ISSUE THREAD

**Issue by Roger-luo** [JuliaQuantum/QuCmp.jl#15]
For disscussion about circuit visualization.

**Comment by acroy** [comment]
Maybe you discussed this already, but [Plots.jl](https://github.com/tbreloff/Plots.jl) might be worth a look. Especially the concept of [recipies](https://juliaplots.github.io/ecosystem/) and the independence from the actual graphics framework are very interesting IMO.

**Comment by quantshah** [comment]
@acroy The concept of recipies looks interesting. I found some more [documentation](https://juliaplots.github.io/recipes/). Should we look for a plotting mechanism for data visualization or should we look at libraries that allow plotting shapes polygons and drawing. Most data visualization libraries probably work with coordinates and datapoints etc. But, for circuits a more generic tool might be helpful in the long run. For example in MATLAB's Simulink tool box you can add components interactively and make circuits by joining blocks. If later QuCmp's circuit simulation library has to develop into something like IBM's quantum circuit simulation tool or even a more interactive tool like Quirk : http://algorithmicassertions.com/2016/05/22/quirk.html a more generic drawing package should help. I especially like the interactive idea in Quirk (http://algorithmicassertions.com/quirk)

**Comment by quantshah** [comment]
Some more engines and libraries from the Gitter discussion :
1. Latex package for circuits : http://physics.unm.edu/CQuIC/Qcircuit/ 
   @i2000s might know about this. If you just want circuit visualization, then a function to write a latex file which visualizaes the circuit based on this package can be developed. This is the same approach used by QuTip here : https://github.com/qutip/qutip/blob/master/qutip/qip/circuit.py#L927. The latex code generated by this function based on the circuit and gates is then compiled by https://github.com/qutip/qutip/blob/master/qutip/qip/circuit_latex.py#L45. It can be directly used since Julia has a [Latex](https://github.com/rened/LaTeX.jl) package with image handling.
2. Using compose.jl : https://github.com/dcjones/Compose.jl
   It can create shapes and polygons easily as you can see in the example.
3. Using the [Cairo plotting](https://github.com/JuliaGraphics/Cairo.jl/blob/master/samples/Samples.md) engine directly. This gives more flexibility and is more like a pixel by pixel drawing of shapes.
4. OpenGl based package GLVisualize. (http://www.glvisualize.com/)
   This gives much more flexibility and will especially be helpful for 3D plotting such as states in a Bloch Sphere or to animate rotation of state vectors etc etc

**Comment by Roger-luo** [comment]
@acroy yes, I just move gitter discussion to this issue, in case we lost some idea... I'll try Plots.jl separately to see if it could acheive the goal of interactivity.

@sahmed95 you are right I'm thinking about interactive visualization for circuit indeed. Quirk is what I imagined for the visualization. JS engine always have a good interactive engine in a browser. I'll try the latex implementation. 

I guess a parser for generating js file for the plots which is similar to qutip's implementation will be a better solution.
